xds/src/main/java/io/grpc/xds/client/XdsResourceType.java

```diff
@@ -148,15 +148,25 @@ ValidatedResourceUpdate<T> parse(Args args, List<Any> resources) {
       Any resource = resources.get(i);
 
       Message unpackedMessage;
+      String wrappedResourceName = null;
       try {
-        resource = maybeUnwrapResources(resource);
+        if (resource.getTypeUrl().equals(TYPE_URL_RESOURCE)) {
+          Resource wrappedResource = unpackCompatibleType(resource, Resource.class,
+              TYPE_URL_RESOURCE, null);
+          resource = wrappedResource.getResource();
+          wrappedResourceName = wrappedResource.getName();
+        } 
         unpackedMessage = unpackCompatibleType(resource, unpackedClassName(), typeUrl(), null);
       } catch (InvalidProtocolBufferException e) {
         errors.add(String.format("%s response Resource index %d - can't decode %s: %s",
                 typeName(), i, unpackedClassName().getSimpleName(), e.getMessage()));
         continue;
       }
-      String name = extractResourceName(unpackedMessage);
+      String name = wrappedResourceName;
```

Comment:
It is sort of strange to assign to `name` here, as you are not doing SSA and you may also immediately rewrite the result but based on other variables. It makes the data flow irregular.

I'd have expected just reusing `name` instead of having `wrappedResourceName`, an if-else here for SSA, or `if (name == null || name.isEmpty()) `.

Reply:
Moved the logic around to simplify this, let me know if this looks better.